### PR TITLE
lockfile optimizer v2

### PR DESCRIPTION
**Pull request by indutny**
Here we go again :joy:

I decided to disable it by default and thankfully the way it is implemented means that we optimize existing lockfile so there are no changes to the resolution process! I also added `--optimize-lockfile` CLI option so we could test it out if needed. It should be fairly easy to rip it out if we ever decide to do it.

### CODE REVIEW

File: crates/node-maintainer/src/maintainer.rs [hunk 1]
```diff
@@ -84,6 +91,9 @@ impl NodeMaintainerOptions {
         nm.run_resolver(self.kdl_lock).await?;
         #[cfg(debug_assertions)]
         nm.graph.validate()?;
```
Comment (indutny) [inline crates/node-maintainer/src/maintainer.rs:93]: Now you might be wondering why don't validate after optimization? Unfortunately the edges in the optimized tree are not quite right. I do clone outgoing edges when duplicating nodes, but they point to original graph nodes and thus sometimes validation fails with "unreachable dependency" error.

File: crates/node-maintainer/src/optimizer.rs [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+            res.push((
+                format!("{}@{}", node.package.name, node.package.version),
+                children,
+            ));
+        }
+        res
+    }
+
+    #[test]
+    fn it_promotes_shared_dependencies() {
```
Comment (indutny) [inline crates/node-maintainer/src/optimizer.rs:454]: We have tests!

**Comment by indutny** [comment]
Pushed one more optimization. It still produces a bit different results on existing maintainer tests, but now the result of optimizer is a bit more shallow (since I retry promoting leafs multiple times to let it settle).

**Comment by indutny** [comment]
FWIW, this doesn't respect peer dependencies, but neither does the non-optimizing resolver at the moment.

**Comment by zkat** [comment]
This is really weird. When I use the optimized lockfile to restore, I get this error:

```
Error: nassun::io::extract

  × Failed to extract tarball. The file exists. (os error 80) (file: .\node_modules\@testing-library/jest-dom/
  │ node_modules/chalk\license)
  ╰─▶ The file exists. (os error 80)
```

(I added the file path reporting in `main`, so you'll have to rebase in order to get that extra info)

The confusing thing is that I'm looking at the lockfile, and I only see one entry for `pkg "@testing-library/jest-dom" "chalk"`, so something's going weird with the lockfile loading...

**Comment by indutny** [comment]
Ah, I think I know what might be going on here. Sorry about that! Will take a look tomorrow.

**Comment by zkat** [comment]
update: this seems to be happening in `main` too, so

**Comment by indutny** [comment]
I fixed the validator, but there are some bugs in the optimization process that I need to squash before this is ready to merge.

**Comment by indutny** [comment]
Blargh, this algorithm is a bust. It can't handle:
```
        ├─┬ @testing-library/react@13.4.0
            │ └─┬ @testing-library/dom@8.20.0
            │   ├─┬ @babel/code-frame@7.18.6
            │   │ └─┬ @babel/highlight@7.18.6
            │   │   └── chalk@2.4.2
            │   └── chalk@4.1.2
```

If there are also a bunch of `chalk@4.1.2` and slightly less `chalk@2.4.2` in surrounding dependencies. It pushes the `chalk@2.4.2` into `testing-library/react` and it breaks `@babel/highlight`.

I'm going to salvage useful parts of this PR, but for now I'm finished with this idea.

**Pull request closed without merging**